sig-and-wg/sig-core/proposals/common-validator.md

```diff
@@ -0,0 +1,266 @@
+
+ After unification, `makefile` for `ui-api-layer` component will look like that:
+
+ ```makefile
+APP_NAME = ui-api-layer
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build test format lint build-image push-image
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
+
+resolve:
+	dep ensure -v -vendor-only
+build:
+	go build -o $(APP_NAME)
+test:
+	go test ./...
+format:
+	go fmt
+lint:
+	go vet
+
+build-image:
+	docker build -t $(APP_NAME) .
+push-image:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
+	docker push $(IMG):$(TAG)
+ ```
+
+### Common Makefile
+
+As in previous point unified targets were introduced it would be fine to not duplicate all that `makefile` content in other components. Fortunately it is possible, because it is possible to include `makefiles`. Thanks to that we can create one common `makefile` that will be included in components `makefiles`.
+
+Lets name this common `makefile` `template.mk` - files with `*.mk` are threated as `makefiles` - and put it in `kyma/common` directory.
+
+The content of `template.mk` file will look like that:
+
+```makefile
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build test format lint build-image push-image
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
+
+resolve:
+	dep ensure -v -vendor-only
+build:
+	go build -o $(APP_NAME)
+test:
+	go test ./...
+format:
+	go fmt
+lint:
+	go vet
+
+build-image:
+	docker build -t $(APP_NAME) .
+push-image:
+	docker tag $(APP_NAME) $(IMG):$(TAG)
+	docker push $(IMG):$(TAG)
+```
+
+And `makefile` for `ui-api-layer`:
+
+```makefile
+APP_NAME = ui-api-layer
+REPOSITORY_PATH = $(realpath $(shell pwd)/../..)
+include $(REPOSITORY_PATH)/common//template.mk
+```
+
+Thanks to that all targets from `template.mk` are available in `makefile` for `ui-api-layer`.
+
+Unfortunately the structure of components may vary and they may be built in different way. For example `ui-api-layer` has `main.go` file on the root of the component, `binding-usage-controller` has it in `cmd/controller` directory. The template should handle such situation and it is also possible thanks to `template definition` and `eval` function.
+
+After introducing `template definition` the `makefiles` will look like that:
+
+`template.mk`:
+```makefile
+IMG = $(DOCKER_PUSH_REPOSITORY)$(DOCKER_PUSH_DIRECTORY)/$(APP_NAME)
+TAG = $(DOCKER_TAG)
+
+.PHONY: ci-pr ci-master ci-release resolve build test format lint build-image push-image
+
+ci-pr: validate build-image push-image
+ci-master: ci-pr
+ci-release: ci-master
+
+validate: resolve build test format lint
+
+define TARGETS
+resolve:
+	dep ensure -v -vendor-only
+build:
+	go build -o $(APP_NAME) $(1)
+test:
+	go test ./...
```

Comment:
We may also provide `-race` flag to `go test` to enable race detection.

Reply:
The main focus of this proposal is architecture, not what is executed in `make` targets. I expect also discussion about `gometalinter`, `goimports` or other tools, but it should be discussed during implementation I think.

Anyway, I will add this flag to the proposal.